### PR TITLE
feat(queue): add more details in case of train reset

### PR DESCRIPTION
**Pull request by sileht**
Fixes MRGFY-547

Change-Id: I656dad5a333bf2b4c7f5db52e9af0f65c6ad5341

### CODE REVIEW

File: mergify_engine/queue/merge_train.py [hunk 1]
```diff
@@ -57,6 +57,34 @@
...
+class UnexpectedChange:
+    pass
+
+
+@dataclasses.dataclass
+class UnexpectedDraftPullRequestChange(UnexpectedChange):
+    draft_pull_request_number: github_types.GitHubPullRequestNumber
+
+    def __str__(self) -> str:
+        return f"the draft pull request {self.draft_pull_request_number}has been manually updated"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:69]:
```suggestion
        return f"the draft pull request #{self.draft_pull_request_number} has been manually updated"
```

File: mergify_engine/queue/merge_train.py [hunk 2]
```diff
@@ -57,6 +57,34 @@
...
+
+
+@dataclasses.dataclass
+class UnexpectedUpdatedPullRequestChange(UnexpectedChange):
+    updated_pull_request_number: github_types.GitHubPullRequestNumber
+
+    def __str__(self) -> str:
+        return f"the updated pull request #{self.updated_pull_request_number} has been manually updated"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:77]: not really clear

File: mergify_engine/engine/queue_runner.py [hunk 3]
```diff
@@ -153,23 +163,23 @@ async def handle(queue_rules: rules.QueueRules, ctxt: context.Context) -> None:
...
         )
-        await train.reset()
+        await train.reset(unexpected_changes)
 
-    if unexpected_changes:
         await ctxt.client.post(
             f"{ctxt.base_url}/issues/{ctxt.pull['number']}/comments",
             json={
-                "body": "This pull request has unexpected changes. The whole train will be reset."
+                "body": f"This pull request has unexpected changes, {unexpected_changes}. The whole train will be reset."
```
Comment (jd) [inline mergify_engine/engine/queue_runner.py:183]:
```suggestion
                "body": f"This pull request has unexpected changes: {unexpected_changes}. The whole train will be reset."
```

File: mergify_engine/queue/merge_train.py [hunk 4]
```diff
@@ -57,6 +57,34 @@
...
+
+
+@dataclasses.dataclass
+class UnexpectedBaseBranchChange(UnexpectedChange):
+    base_sha: github_types.SHAType
+
+    def __str__(self) -> str:
+        return f"the base branch head has moved manually to {self.base_sha}"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:85]:
```suggestion
        return f"an external action moved the branch head {self.base_sha}"
```
Not a fan of this, but I think it sounds better as it really indicates something which is not Mergify did something wrong.
Better idea?

File: mergify_engine/queue/merge_train.py [hunk 5]
```diff
@@ -687,8 +715,8 @@ async def update_summaries(
                 )
                 show_queue = False
             elif conclusion == check_api.Conclusion.PENDING:
-                if will_be_reset:
-                    headline = f"✨ Unexpected queue change. The pull request {self._get_user_refs()} will be re-embarked soon. ✨"
+                if unexpected_change is not None:
+                    headline = f"✨ Unexpected queue change, {unexpected_change}. The pull request {self._get_user_refs()} will be re-embarked soon. ✨"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:719]:
```suggestion
                    headline = f"✨ Unexpected queue change: {unexpected_change}. The pull request {self._get_user_refs()} will be re-embarked soon. ✨"
```

File: mergify_engine/queue/merge_train.py [hunk 6]
```diff
@@ -773,20 +801,25 @@ async def update_summaries(
...
             if conclusion == check_api.Conclusion.SUCCESS:
                 original_pull_title = f"The pull request embarked with {self._get_embarked_refs(include_my_self=False)} is mergeable"
             elif conclusion == check_api.Conclusion.PENDING:
                 original_pull_title = f"The pull request is embarked with {self._get_embarked_refs(include_my_self=False)} for merge"
             else:
                 original_pull_title = f"The pull request embarked with {self._get_embarked_refs(include_my_self=False)} cannot be merged and has been disembarked"
+        else:
+            original_pull_title = "The pull request is going to be re-embarked soon"
+            unexpected_change_summary = (
+                f"✨ Unexpected queue change, {unexpected_change}. ✨\n\n"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:815]:
```suggestion
                f"✨ Unexpected queue change: {unexpected_change}. ✨\n\n"
```

File: mergify_engine/queue/merge_train.py [hunk 7]
```diff
@@ -1011,8 +1044,10 @@ async def _sync_configuration_change(self, queue_rules: rules.QueueRules) -> Non
                     i, reason="The associated queue rule does not exist anymore"
                 )
 
-    async def reset(self) -> None:
-        await self._slice_cars(0, reason="Unexpected queue change")
+    async def reset(self, unexpected_change: UnexpectedChange) -> None:
+        await self._slice_cars(
+            0, reason=f"Unexpected queue change, {unexpected_change}."
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:1049]:
```suggestion
            0, reason=f"Unexpected queue change: {unexpected_change}."
```

**Review by jd — CHANGES_REQUESTED** [review]

**Comment by sileht** [comment]
@Mergifyio squash first-commit

**Review by jd — APPROVED** [review]

**Comment by jd** [comment]
@mergifyio refresh

**Pull request merged**